javapoet/src/main/java/com/palantir/javapoet/TypeSpec.java

```diff
@@ -913,6 +958,18 @@ public TypeSpec build() {
                 }
             }
 
+            if (kind == Kind.RECORD) {
+                checkState(!modifiers.contains(Modifier.ABSTRACT), "abstract record");
+            }
+
+            if (recordConstructor != null && !recordConstructor.parameters().isEmpty()) {
+                checkState(kind == Kind.RECORD, "%s is not record", this.name);
+                for (ParameterSpec recordComponent : recordConstructor.parameters()) {
+                    checkState(recordComponent != null, "recordComponents contain null");
+                    checkState(recordComponent.modifiers().isEmpty(), "recordComponents has modifier");
+                }
+            }
```

Comment:
Couple things:
- The second condition here is unnecessary
- The kind is already validated by the builder method
- It's not the responsibility of `TypeSpec` to check if a method has null parameters
- Use `checkArgument` instead of `checkState`

```suggestion
            if (recordConstructor != null) {
                for (ParameterSpec recordComponent : recordConstructor.parameters()) {
                    checkArgument(recordComponent.modifiers().isEmpty(), "record components must not have modifiers");
                }
            }
```